### PR TITLE
Notifications - Step 1

### LINKED ISSUE #145: Complete notifications system
Today Gitea can send e-mail for new Issues/PRs/Comments and that's it.

I think it should have a complete notifications system, like GitHub and GitLab.

- [x] Notifications would be saved in a database table
- [x] A page to manage them. Equivalent to https://github.com/notifications
- [x] Mark as read/unread, pinning a notification (visible even if read)
- [x] Subscribe/Unsubscribe button
- [x] Mark all notifications as read
- [ ] Mark all notifications as unread
- [ ] Automatic subscribe on Open/Comment, etc
- [ ] Also more refined watch options?
- [ ] E-mail read means notification gone. Each user should have the ability to disable this
- [ ] E-Mail notifications for pushes

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/39150685-complete-notifications-system?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by andreynering**
First step to resolve #145

This deprecates #321

I plan 2 steps:

- [ ] Create database table and create/update notifications (this one)
- [ ] Actually have UI for notifications

### CODE REVIEW

File: models/notification.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+)
+
+const (
+	// NotificationStatusUnread represents an unread notification
+	NotificationStatusUnread NotificationStatus = iota + 1
+	// NotificationStatusRead represents a read notification
+	NotificationStatusRead
+)
+
+const (
```
Comment (tboerger) [inline models/notification.go:21]: Just put the constants in a single block

File: models/notification.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+	NotificationSourceCommit
+)
+
+// Notification represents a notification
+type Notification struct {
+	ID     int64 `xorm:"pk autoincr"`
+	UserID int64 `xorm:"INDEX NOT NULL"`
+	RepoID int64 `xorm:"INDEX NOT NULL"`
+
+	Status NotificationStatus `xorm:"SMALLINT INDEX NOT NULL"`
```
Comment (tboerger) [inline models/notification.go:36]: Really a separate type for simple integers?

**Comment by tboerger** [comment]
Generally not bad, but you should select more collections from the db instead of firing lots of small queries

File: models/notification.go (hunk 2 above)
Comment (lunny) [reply models/notification.go:36]: I really like this. :)

File: models/notification.go (hunk 1 above)
Comment (andreynering) [reply models/notification.go:21]: It's necessary for iota and different types, as discussed below.

File: models/notification.go (hunk 2 above)
Comment (andreynering) [reply models/notification.go:36]: I also prefer enum, and others too, since I was asked to use this in the previous PR.

**Comment by andreynering** [comment]
> you should select more collections from the db instead of firing lots of small queries

Updated. No more queries inside the loop.

File: models/notification.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+// CreateOrUpdateIssueNotifications creates an issue notification
+// for each watcher, or updates it if already exists
+func CreateOrUpdateIssueNotifications(issue *Issue, notificationAuthorID int64) error {
+	sess := x.NewSession()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+	defer sess.Close()
```
Comment (lunny) [inline models/notification.go:82]: put `sess.Close()` before `sess.Begin()`

**Review by lunny — CHANGES_REQUESTED** [review]
Except that then LGTM

**Comment by andreynering** [comment]
Deprecated by https://github.com/go-gitea/gitea/pull/523

**Pull request closed without merging**